samples/net/cellular_modem/src/main.c

```diff
@@ -0,0 +1,245 @@
+	}
+}
+
+static void sample_dns_request_result(enum dns_resolve_status status, struct dns_addrinfo *info,
+				      void *user_data)
+{
+	if (sample_test_dns_in_progress == false) {
+		return;
+	}
+
+	if (status != DNS_EAI_INPROGRESS) {
+		return;
+	}
+
+	sample_test_dns_in_progress = false;
+
+	sample_test_dns_addrinfo = *info;
+
+	k_sem_give(&dns_query_sem);
+}
+
+static int sample_dns_request(void)
+{
+	static uint16_t dns_id;
+	int ret;
+
+	sample_test_dns_in_progress = true;
+
+	ret = dns_get_addr_info(SAMPLE_TEST_ENDPOINT_HOSTNAME,
+				DNS_QUERY_TYPE_A,
+				&dns_id,
+				sample_dns_request_result,
+				NULL,
+				19000);
+	if (ret < 0) {
+		return -EAGAIN;
+	}
+
+	if (k_sem_take(&dns_query_sem, K_SECONDS(20)) < 0) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+int sample_echo_packet(struct sockaddr *ai_addr, socklen_t ai_addrlen)
+{
+	int ret;
+	int socket_fd;
+
+	printk("Opening UDP socket\n");
+
+	socket_fd = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+
+	if (socket_fd < 0) {
+		printk("Failed to open socket\n");
+		return -1;
+	}
+
+	printk("Socket opened\n");
+
+	if (ai_addr->sa_family == AF_INET) {
+		net_sin(ai_addr)->sin_port = htons(SAMPLE_TEST_ENDPOINT_UDP_ECHO_PORT);
+	} else if (ai_addr->sa_family == AF_INET6) {
+		net_sin6(ai_addr)->sin6_port = htons(SAMPLE_TEST_ENDPOINT_UDP_ECHO_PORT);
+	} else {
+		printk("Unsupported address family\n");
+		return -1;
+	}
+
+	printk("Sending sample test packet\n");
+
+	ret = zsock_sendto(socket_fd, sample_test_packet, sizeof(sample_test_packet), 0,
+			   ai_addr, ai_addrlen);
+
+	if (ret < sizeof(sample_test_packet)) {
+		printk("Failed to send sample test packet\n");
+		return -1;
+	}
+
+	printk("Await sample test packet echo\n");
+
+	ret = zsock_recv(socket_fd, sample_recv_buffer, sizeof(sample_recv_buffer), 0);
+
+	if (ret != sizeof(sample_test_packet)) {
+		printk("Echoed sample test packet has incorrect size\n");
+		return -1;
+	}
+
+	if (memcmp(sample_test_packet, sample_recv_buffer, sizeof(sample_recv_buffer)) != 0) {
+		printk("Echoed sample test packet data mismatch\n");
+		return -1;
+	}
+
+	printk("Echo recevied successfully!\n");
+
+	printk("Close UDP socket\n");
+
+	ret = zsock_close(socket_fd);
+
+	if (ret < 0) {
+		printk("Failed to close socket\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+int main(void)
+{
+	uint32_t raised_event;
+	const void *info;
+	size_t info_len;
+	int ret;
+
+	init_sample_test_packet();
+
+	printk("Powering on modem\n");
+
+	pm_device_action_run(modem, PM_DEVICE_ACTION_RESUME);
+
+	printk("Bring up network interface\n");
+
+	ret = net_if_up(net_if_get_default());
+
+	if (ret < 0) {
+		printk("Failed to bring up network interface\n");
+		return -1;
+	}
+
+	printk("Waiting for L4 connected\n");
+
+	ret = net_mgmt_event_wait_on_iface(net_if_get_default(),
+					   NET_EVENT_L4_CONNECTED, &raised_event, &info,
+					   &info_len, K_SECONDS(120));
+
+	if (ret != 0) {
+		printk("L4 was not connected in time\n");
+		return -1;
+	}
```

Comment:
This doesn't seem to be sufficient for all cases.
I tested the sample and this event fires before an IPv4 address is ready on the PPP network interface. Then the DNS resolution fails.
If a delay is added here ~30 seconds, then the DNS resolution succeeds because the interface is fully ready.

I tried adding the following but it doesn't seem to work quite right, the 30 second timeout always happens instead of receiving the address add event:
```
	printk("Waiting for IP Address\n");

	ret = net_mgmt_event_wait_on_iface(net_if_get_default(),
					   NET_EVENT_IPV4_CMD_ADDR_ADD, &raised_event, &info,
					   &info_len, K_SECONDS(30));
	if (ret != 0) {
		printk("Timeout waiting for IP address, continue anyway\n");
	}
```

Reply:
The sample automatically sets up the DNS server based on the provided name servers. The DNS relies on L4 to send IP packets, so as soon as L4 is ready (which is after L2 is set up and IPCP has negotiated IP addresses) the DNS request should succeed.

The root issue in your case seems to be that you are getting a spurious NET_EVENT_L4_CONNECTED event for some reason, maybe something in the driver is exploding after the event occurred causing the network to fail? I really don't understand how L4=Connected can occur before the network is ready

Reply:
I see `connection_manager` is experimental. Maybe there is a bug? I can provide the debug logs and you can see exactly the event order... Not sure if you care to look or not.

Reply:
See lines 340-358. Then line 432.
[run_23-07-17_1.log](https://github.com/zephyrproject-rtos/zephyr/files/12073791/run_23-07-17_1.log)